tests/aws/services/stepfunctions/v2/callback/test_callback.py

```diff
@@ -471,3 +511,141 @@ def test_start_execution_sync_delegate_timeout(
+    @markers.aws.needs_fixing
+    def test_multiple_executions_and_heartbeat_notifications(
+        self,
+        aws_client,
+        create_iam_role_for_sfn,
+        create_state_machine,
+        sqs_create_queue,
+        sfn_snapshot,
+    ):
+        sfn_snapshot.add_transformer(sfn_snapshot.transform.sqs_api())
+        sfn_snapshot.add_transformer(
+            JsonpathTransformer(
+                jsonpath="$..TaskToken",
+                replacement="a_task_token",
+                replace_reference=False,
+            )
+        )
+        sfn_snapshot.add_transformer(
+            JsonpathTransformer(
+                jsonpath="$..MessageId",
+                replacement="a_message_id",
+                replace_reference=False,
+            )
+        )
+
+        queue_name = f"queue-{short_uid()}"
+        queue_url = sqs_create_queue(QueueName=queue_name)
+        sfn_snapshot.add_transformer(RegexTransformer(queue_url, "sqs_queue_url"))
+        sfn_snapshot.add_transformer(RegexTransformer(queue_name, "sqs_queue_name"))
+
+        sfn_role_arn = create_iam_role_for_sfn()
+
+        template = CT.load_sfn_template(
+            TT.SERVICE_SQS_SEND_AND_WAIT_FOR_TASK_TOKEN_WITH_HEARTBEAT_PATH
+        )
+        definition = json.dumps(template)
+
+        creation_response = create_state_machine(
+            name=f"state_machine_{short_uid()}", definition=definition, roleArn=sfn_role_arn
+        )
+        sfn_snapshot.add_transformer(sfn_snapshot.transform.sfn_sm_create_arn(creation_response, 0))
+        state_machine_arn = creation_response["stateMachineArn"]
+
+        exec_input = json.dumps(
+            {"QueueUrl": queue_url, "Message": "txt", "HeartbeatSecondsPath": 60}
+        )
+
+        # Launch multiple execution of the same state machine.
+        execution_count = 6
+        execution_arns = list()
+        for _ in range(execution_count):
+            execution_arn = aws_client.stepfunctions.start_execution(
+                stateMachineArn=state_machine_arn, input=exec_input
+            )["executionArn"]
+            execution_arns.append(execution_arn)
+
+        # Launch one sqs task token handler per each execution, and await for all the terminate handling the task.
+        task_token_handler_latch = CountDownLatch(execution_count)
+
+        def _sqs_task_token_handler(inner_handler_id: int):
+            task_token_handler_responses = _handle_sqs_task_token_with_heartbeats_and_success(
+                aws_client, queue_url
+            )
+            sfn_snapshot.match(
+                f"task_token_inner_handler_{inner_handler_id}_responses",
+                task_token_handler_responses,
+            )
+            task_token_handler_latch.count_down()
+
+        for i in range(execution_count):
+            inner_handler_thread = threading.Thread(target=_sqs_task_token_handler, args=(i,))
+            inner_handler_thread.start()
```

Comment:
good to know 💡 : if you ever need a scenario where multiple threads should be ready and then start at the same time, a barrier could be useful. Example in Lambda perftests: https://github.com/localstack/localstack/blob/0f4b52eefee5b9eca65c2050cd69ec423e3eea9f/tests/aws/services/lambda_/test_lambda_performance.py#L302

Reply:
Thank you! -- In this case we don't need for all threads to wait on a barrier as they interface with separate state machines